ra/registration-authority.go

```diff
@@ -480,6 +478,77 @@ func (ra *RegistrationAuthorityImpl) NewCertificate(req core.CertificateRequest,
 	return cert, nil
 }
 
+// domainsForRateLimiting transforms a list of FQDNs into a list of eTLD+1's
+// for the purpose of rate limiting. It also de-duplicates the output
+// domains.
+func domainsForRateLimiting(names []string) ([]string, error) {
+	domainsMap := make(map[string]struct{}, len(names))
+	var domains []string
+	for _, name := range names {
+		eTLDPlusOne, err := publicsuffix.EffectiveTLDPlusOne(name)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := domainsMap[eTLDPlusOne]; !ok {
+			domainsMap[eTLDPlusOne] = struct{}{}
+			domains = append(domains, eTLDPlusOne)
+		}
+	}
+	return domains, nil
+}
+
+func (ra *RegistrationAuthorityImpl) checkCertificatesPerNameLimit(names []string, limit cmd.RateLimitPolicy) error {
+	names, err := domainsForRateLimiting(names)
+	if err != nil {
+		return err
+	}
+	now := ra.clk.Now()
+	windowBegin := limit.WindowBegin(now)
+	counts, err := ra.SA.CountCertificatesByNames(names, windowBegin, now)
+	if err != nil {
+		return err
+	}
+	var badNames []string
+	for _, name := range names {
+		count, ok := counts[name]
+		if !ok {
+			// Shouldn't happen, but let's be careful anyhow.
+			return errors.New("StorageAuthority failed to return a count for every name")
+		}
+		if count >= int(limit.GetThreshold(name)) {
```

Comment:
Ah, I think we have other places where we rely on being run on an 64-bit machine, but this one is subtle. Could we do int64(count), instead?